client/rpc-servers/src/middleware.rs

```diff
@@ -51,22 +79,45 @@ impl RpcMetrics {
 }
 
 /// Middleware for RPC calls
-pub struct RpcMiddleware {
+pub struct RpcMiddleware<M: Metadata, CM: CustomMiddleware<M> = ()> {
 	metrics: RpcMetrics,
 	transport_label: String,
+	custom_middleware: CM,
```

Comment:
I already thought about it at the very beginning but we can't do that because the trait needs a generic for the Metadata. Or we should define the Metadata type in sc-rpc-server, or sc-rpc-server depends on sc-rpc-api.

Reply:
We can pass the generic parameter here? What is the problem?

Reply:
No, Rust doesn't allow that: https://doc.rust-lang.org/error-index.html#method-has-generic-type-parameters

Reply:
Because:

> However, with trait objects we have to make a table containing every object that implements the trait. Now, if it has type parameters, we need to add implementations for every type that implements the trait, and there could theoretically be an infinite number of types.

Reply:
@bkchr also dynamic typing induces a slight overhead because of the vtable, in most contexts it's not annoying, but in this context in particular it's annoying, because we're talking about methods that are called at each rpc request.
So I think it is important to use static typing for the rpc middleware.

Reply:
I know how rust works 😉

The Metadata is fixed anyway from the beginning we use it in sc_service.

> I already thought about it at the very beginning but we can't do that because the trait needs a generic for the Metadata. Or we should define the Metadata type in sc-rpc-server, or sc-rpc-server depends on sc-rpc-api. 

You still pass the Metadata generic, as I illustrated it above.

Reply:
Elois is right. Since the `CustomMiddleware` trait contains functions like `fn on_request<F, X>` which have type parameters, the trait is not **object-safe** and cannot be used for a **trait objects**.

The documentation linked by Elois explains well this problem. However it suggests having a `where Self: Sized` bound on the function, but I never did that and don't know what are the limitations.

Reply:
You are right, but I think Basti is suggesting a different approach, maybe a Trait Bound. I've talked to Elois and we will see what we can do

Reply:
I have just re-investigated in detail and unfortunately I confirm that it is indeed impossible to use a dyn trait in the case of the CustomMiddleware trait, but I had misunderstood last month the exact causes, so my explanations were partly incorrect.
The real reason why the CustomMiddleware trait can't be a dyn trait, we say in Rust vocabulary that it's not "object safe", is because its methods (on_request and on_call) take generic types ("type parameters" in Rust vocabulary), which makes these methods non-dispatchable from a dyn context.

> However it suggests having a where Self: Sized bound on the function

This suggestion of the doc allows to make a trait object safe even if some of its methods are not, the `where Self: Sized` bound makes it explicit which methods are not dispatchable. But the constraint is that the methods in question cannot be called from a generic context, so in our case we would end up with a trait from which no method can be called, so this is useless in our case.

> I know how rust works wink

No one can claim to master perfectly all the workings of Rust except maybe some members of the core Rust team, and even then! I've been practicing Rust daily for more than 4 years and I'm still learning from time to time.
@bkchr, please take the time to read the link to the official Rust doc that explains why what you are asking is impossible, I am not a beginner, if there was a way I would have found it a long time ago given my experience in Rust.

https://doc.rust-lang.org/error-index.html#method-has-generic-type-parameters

Reply:
@bkchr This reference lists the constraints that a trait must satisfy to be used in a dyn context: https://doc.rust-lang.org/reference/items/traits.html#object-safety

The `CustonMiddleware` trait does not respect the following constraint:

> All associated functions must either be dispatchable from a trait object or be explicitly non-dispatchable

The 2 methods of the `CustonMiddleware` trait (`on_request` and `on_call`) are not dispatchable because they do not respect the following constraint: 

> Not have any type parameters (although lifetime parameters are allowed),

We can make these 2 methods explicitly non-dispatchable, but then we can't call them from a dyn context, and we need to call them, so it doesn't solve the problem.
We can't remove type parameters from these methods because they are called inside methods of the same name of the `jsonrpc_core::Middleware` trait, a trait that requires type parameters at the method level.